### ISSUE TITLE
Investigate long linking times

### ISSUE THREAD

**Issue by rudolfs** [radicle-dev/radicle-upstream#444]
Linking the proxy binary takes 3m on my 2015 mbp, we lose a lot of time waiting.
Look if there is any way to reduce the linking time.

Timebox: 1 day.

```
Machine
-------
15-inch Mid 2015 mbp
2,8GHz Quad-Core Intel Core i7
16GB RAM


Baseline 
--------
Finished dev [unoptimized + debuginfo] target(s) in 12m 10s  // compile from scratch
Finished dev [unoptimized + debuginfo] target(s) in 12m 09s  // compile from scratch
Finished dev [unoptimized + debuginfo] target(s) in 3m 01s   // link-only
Finished dev [unoptimized + debuginfo] target(s) in 3m 03s   // link-only

debug = false
=============
Finished dev [unoptimized] target(s) in 11m 18s  // compile from scratch
Finished dev [unoptimized] target(s) in 2m 57s   // link-only


nightly-2020-06-03
==================
Finished dev [unoptimized] target(s) in 12m 19s  // compile from scratch
Finished dev [unoptimized] target(s) in 3m 41s   // link-only


nightly-2020-06-03 + CARGO_BUILD_PIPELINING=true
==================
Finished dev [unoptimized] target(s) in 12m 03s  // compile from scratch
Finished dev [unoptimized] target(s) in 3m 43s   // link-only
```

Profiling info via `cargo build -Ztimings`:

<img width="1179" alt="Screenshot 2020-06-04 at 00 28 38" src="https://user-images.githubusercontent.com/158411/83695547-921cf600-a5fa-11ea-97a4-e8aedb29a6ef.png">

<img width="1167" alt="Screenshot 2020-06-04 at 00 28 42" src="https://user-images.githubusercontent.com/158411/83695550-95b07d00-a5fa-11ea-9eb5-029d98082f41.png">

```
    |  Unit                                     | Total  | Codegen     | Features
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1.  | proxy v0.1.0 bin "proxy"                  | 509.5s | 0.0s (0%)   |
2.  | libsodium-sys v0.2.5 custom-build (run)   | 122.7s | 0.0s (0%)   |
3.  | nalgebra v0.18.1                          | 68.1s  | 0.3s (0%)   | default, matrixmultiply, std
4.  | libgit2-sys v0.9.2 custom-build (run)     | 66.9s  | 0.0s (0%)   | https, libssh2-sys, openssl-sys, ssh, ssh_key_from_memory
5.  | libssh2-sys v0.2.16 custom-build (run)    | 54.8s  | 0.0s (0%)   |
6.  | backtrace-sys v0.1.35 custom-build (run)  | 54.4s  | 0.0s (0%)   | backtrace-sys
7.  | bzip2-sys v0.1.8+1.0.8 custom-build (run) | 54.1s  | 0.0s (0%)   |
8.  | nettle-sys v2.0.4 custom-build (run)      | 53.9s  | 0.0s (0%)   | nettle-src, vendored
9.  | bindgen v0.51.1                           | 42.3s  | 13.4s (32%) |
10. | lalrpop v0.17.2                           | 41.6s  | 20.8s (50%) |
11. | serde v1.0.106                            | 30.4s  | 0.7s (2%)   | default, derive, serde_derive, std
12. | protobuf v2.14.0                          | 28.1s  | 17.8s (63%) |
13. | sled v0.31.0                              | 27.7s  | 15.9s (57%) | default, no_metrics
14. | sp-api-proc-macro v2.0.0-alpha.5          | 27.2s  | 0.0s (0%)   | default, std
15. | clang-sys v0.28.1                         | 23.3s  | 18.5s (79%) | clang_6_0, gte_clang_3_6, gte_clang_3_7, gte_clang_3_8, gte_clang_3_9, gte_clang_4_0, gte_clang_5_0, gte_clang_6_0, libloading, runtime
16. | syn v1.0.17                               | 21.3s  | 4.5s (21%)  | clone-impls, default, derive, extra-traits, fold, full, parsing, printing, proc-macro, quote, visit, visit-mut
17. | serde_derive v1.0.106                     | 20.0s  | 0.0s (0%)   | default
18. | parity-scale-codec v1.3.0                 | 18.5s  | 0.4s (2%)   | bitvec, default, derive, full, parity-scale-codec-derive, serde, std
19. | libsecp256k1 v0.3.5                       | 17.2s  | 1.4s (8%)   | default, hmac, hmac-drbg, sha2, std, typenum
20. | warp v0.2.2                               | 15.5s  | 2.2s (14%)  | default, indexmap, multipart, openapi, openapiv3, tokio-tungstenite, websocket
...
```
[cargo-timing-20200603T220512Z.html.zip](https://github.com/radicle-dev/radicle-upstream/files/4726521/cargo-timing-20200603T220512Z.html.zip)

**Comment by NunoAlexandre** [comment]
@rudolfs how did you build this cool visualization?

**Comment by rudolfs** [comment]
@NunoAlexandre `cargo build -Ztimings`

**Comment by NunoAlexandre** [comment]
@rudolfs I had run `RUSTFLAGS=-Ztime-passes cargo build` and thought that command above was essentially the same and that you had used some extra tool to get these visualizations. Thanks!

**Comment by rudolfs** [comment]
@NunoAlexandre I overheard yesterday that the linking times could be influenced by pulling in different versions of the same dependencies for both our protocols (coco, registry).

`cargo tree -d` can help shedding some light into it along with [cargo-guppy](https://github.com/facebookincubator/cargo-guppy).
Also [cargo-deny](https://github.com/EmbarkStudios/cargo-deny) could help preventing duplicates from creeping in once if we manage to clean them up.

**Comment by cloudhead** [comment]
Closed in #541
